File: android/app/src/main/kotlin/run/nuri/getagrip/ui/share/RoutineShareSheet.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.share

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.outlined.ContentCopy
import androidx.compose.material.icons.outlined.Share
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import run.nuri.getagrip.engine.FingerSet
import run.nuri.getagrip.engine.L10n
import run.nuri.getagrip.engine.PlanMath
import run.nuri.getagrip.ui.components.EdgeMark
import run.nuri.getagrip.ui.components.SecondaryButton
import run.nuri.getagrip.ui.components.tint
import run.nuri.getagrip.ui.l10n.tr
import run.nuri.getagrip.ui.theme.LocalGripPalette
import run.nuri.getagrip.ui.theme.Metrics

/// A frozen share request, like `ShareCalendarRequest`: the deck can keep moving and the sheet
/// observes NO store, so the code and the name beside it describe one thing.
data class RoutineShareRequest(
    val name: String,
    val metaLine: String,
    val signatureFingers: FingerSet?,
    /// The card's own rung colour: bleu means "nothing resolves", so a red max-day card must not
    /// turn bleu here.
    val peakIntensity: Double?,
    val url: String,
)

/// The routine as a QR code. The code IS the routine — no server, no account, no link to rot —
/// and percentage targets resolve against THEIR maxes.
///
/// TRANSLATION NOTE: iOS shares a rendered PNG via `ShareLink` (AirDrop between phones in a
/// room). Android shares the **LINK** (`ACTION_SEND`, `text/plain`): pasted into any messenger
/// it opens the app via the intent filter, where a PNG of a QR would need a second phone. The
/// QR stays for the in-the-room case. A routine too large for a code (4 KB format vs a level-M
/// QR's capacity) still shares as a link.
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RoutineShareSheet(request: RoutineShareRequest, onClose: () -> Unit) {
    val palette = LocalGripPalette.current
    val context = LocalContext.current
    val haptics = LocalHapticFeedback.current
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false)

    var copied by remember { mutableStateOf(false) }
    // Keyed to the link, so a second presentation never opens on a stale "Link copied".
    LaunchedEffect(request.url) { copied = false }

    ModalBottomSheet(
        onDismissRequest = onClose,
        sheetState = sheetState,
        containerColor = palette.field,
        shape = RoundedCornerShape(topStart = Metrics.radiusSheet, topEnd = Metrics.radiusSheet),
    ) {
        Column(
            Modifier
                .verticalScroll(rememberScrollState())
                .padding(horizontal = Metrics.hPadding)
                .padding(bottom = Metrics.spacing),
            verticalArrangement = Arrangement.spacedBy(18.dp),
        ) {
            Text(
                tr("Share routine"),
                style = MaterialTheme.typography.titleLarge,
                color = palette.inkPrimary,
            )

            // The card's own anatomy (mark, name, plan line), so the sheet reads as the card it came
            // from. One TalkBack stop; read outside the non-composable semantics lambda.
            val spoken = tr("%s. %s", request.name, request.metaLine)
            Column(
                Modifier.clearAndSetSemantics {
                    contentDescription = spoken
                },
                verticalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(10.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    EdgeMark(
                        fingers = request.signatureFingers ?: FingerSet.four,
                        rungTint = PlanMath.IntensityBand.band(request.peakIntensity).tint(palette),
                    )
                    Text(
                        request.name,
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.SemiBold,
                        color = palette.inkPrimary,
                        maxLines = 2,
                    )
                }
                Text(
                    request.metaLine,
                    style = MaterialTheme.typography.bodySmall.copy(fontFeatureSettings = "tnum"),
                    color = palette.inkSecondary,
                )
            }

            Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                QrCodeView(payload = request.url)
            }

            Text(
                tr("To add %s, scan with Camera on iPhone. On Android, open Get a Grip and choose Scan a routine in the Today menu.", request.name),
                style = MaterialTheme.typography.bodySmall,
                color = palette.inkSecondary,
            )

            // Two equal doors: the code for a phone in the room, the link for one that is not. Which you
            // want depends on where the other person is, so neither is filled.
            Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                SecondaryButton(
                    title = if (copied) tr("Link copied") else tr("Copy link"),
                    icon = if (copied) Icons.Filled.Check else Icons.Outlined.ContentCopy,
                    modifier = Modifier.weight(1f),
                ) {
                    if (copyLink(context, request.name, request.url)) {
                        copied = true
                        // The haptic names its cause: the copy LANDED. Clipboard writes can fail (device policy,
                        // stripped ROM), and a buzz over an unchanged clipboard is a small lie.
                        haptics.performHapticFeedback(HapticFeedbackType.Confirm)
                    }
                }
                SecondaryButton(
                    title = tr("Share"),
                    icon = Icons.Outlined.Share,
                    modifier = Modifier.weight(1f),
                ) {
                    shareLink(context, request.name, request.url)
                }
            }

            Spacer(Modifier.padding(bottom = 4.dp))
        }
    }
}

/// The link on the clipboard, labelled with the routine's name, since a clipboard manager
/// shows the label and "getagrip://routine#H4sIA…" is unrecognisable.
///
/// Android 13+ confirms copies itself; the in-app receipt stays because OEM skins vary and the
/// button state is what a screen reader hears.
private fun copyLink(context: Context, name: String, url: String): Boolean = runCatching {
    val clipboard = context.getSystemService(ClipboardManager::class.java) ?: return false
    clipboard.setPrimaryClip(ClipData.newPlainText(name, url))
    true
}.getOrDefault(false)

/// `ACTION_SEND` with `text/plain` — a LINK, not an image (see the header). `EXTRA_SUBJECT`
/// titles emails and some messengers, or the routine arrives as a bare scheme URL.
private fun shareLink(context: Context, name: String, url: String) {
    runCatching {
        val send = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_SUBJECT, L10n.tr("Get a Grip — %s", name))
            putExtra(Intent.EXTRA_TEXT, url)
        }
        context.startActivity(Intent.createChooser(send, null))
    }
}

File: Sources/UI/RoutineShareSheet.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import CoreTransferable
import Foundation
import SwiftUI
import UniformTypeIdentifiers
import UIKit

/// A frozen share request, exactly like `ShareCalendarRequest`. The deck behind the sheet
/// can keep moving — a swipe, a CloudKit merge, an edit landing — and what is on screen
/// stays the routine whose menu was tapped. The sheet observes NO store for the same
/// reason: the code it draws and the name beside it must describe one thing.
struct RoutineShareRequest: Identifiable {
    let id = UUID()
    let name: String
    let metaLine: String
    let signatureFingers: FingerSet?
    /// Carried so the mark here wears the SAME rung colour as the card it was opened
    /// from — bleu specifically means "nothing resolves", and a max-day routine whose
    /// card burns red must not turn bleu one presentation later.
    let peakIntensity: Double?
    let url: URL
}

/// The routine as a QR code. The code IS the routine — there is no server, no account and
/// no link that can rot: everything the recipient's app needs rides in the payload, and
/// percentage targets resolve against THEIR maxes, which is the point of prescribing a
/// fraction rather than a kilogram.
struct RoutineShareSheet: View {
    let request: RoutineShareRequest
    var onClose: () -> Void

    @State private var rendered: RenderedRoutineShare?
    /// The one render attempt failed — terminal for this presentation, and said in
    /// words: a permanently disabled "Preparing image…" is a spinner that lies.
    @State private var renderFailed = false

    private var payload: String { request.url.absoluteString }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 18) {
                    header

                    QRCodeView(string: payload)
                        .frame(maxWidth: .infinity)

                    Text("To add \(request.name), scan with Camera on iPhone. On Android, open Get a Grip and choose Scan a routine in the Today menu.")
                        .font(.system(.footnote))
                        .foregroundStyle(Ink.secondary)
                        .fixedSize(horizontal: false, vertical: true)
                }
                .padding(.horizontal, Metrics.hPadding)
                .padding(.top, 12)
                .padding(.bottom, 12)
                .frame(maxWidth: Metrics.maxContentWidth)
                .frame(maxWidth: .infinity)
            }
            .scrollBounceBehavior(.basedOnSize)
            .scrollEdgeEffectStyle(.soft, for: .bottom)
            .background { AppBackground() }
            .navigationTitle("Share routine")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Close") { onClose() }
                }
            }
            // In the safe area, not at the foot of the scroll — the import sheet's own
            // rule, learned here the measured way: at the `.medium` detent this sheet
            // opens on, an in-document button sat 33 pt into the home-indicator strip,
            // and dragging the content resized the sheet instead of scrolling to it.
            .safeAreaInset(edge: .bottom) { shareAction }
        }
        .presentationDetents([.medium, .large])
        .task {
            if rendered == nil, !renderFailed { await renderImage() }
        }
    }

    /// The card's own anatomy — mark, name, plan line — so the sheet reads as the card it
    /// was opened from rather than as a second description of the same routine.
    private var header: some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 10) {
                EdgeMark(fingers: request.signatureFingers ?? .four,
                         rungTint: PlanMath.IntensityBand.band(for: request.peakIntensity).tint)
                Text(request.name)
                    .font(.system(.title2, weight: .semibold))
                    .foregroundStyle(Ink.primary)
                    .lineLimit(2)
                    .minimumScaleFactor(0.8)
                Spacer(minLength: 0)
            }
            Text(request.metaLine)
                .font(.system(.footnote))
                .monospacedDigit()
                .foregroundStyle(Ink.secondary)
                .lineLimit(2)
                .minimumScaleFactor(0.85)
        }
        // One VoiceOver stop, like the import sheet's twin of this header — same
        // content, same shape under the rotor.
        .accessibilityElement(children: .combine)
    }

    /// Glass is allowed here: a sheet is its own presentation and carries no
    /// `.contextMenu`, so nothing in this tree is ever lifted out from under it.
    @ViewBuilder
    private var shareAction: some View {
        Group {
            if let rendered {
                ShareLink(
                    item: rendered.file,
                    preview: SharePreview(
                        "Get a Grip — \(request.name)",
                        image: Image(uiImage: rendered.image))
                ) {
                    actionLabel(String(localized: "Share the code"), systemImage: "square.and.arrow.up",
                                prominent: true)
                }
                .buttonStyle(.glassProminent)
                .tint(Accent.graphite)
            } else if renderFailed {
                // The on-screen QR above still works — only the PNG could not be baked,
                // and a sentence beats a control that looks tappable and never will be.
                Text("Couldn't prepare a shareable image. The code above still scans.")
                    .font(.system(.footnote, weight: .medium))
                    .foregroundStyle(Ink.secondary)
                    .multilineTextAlignment(.center)
                    .fixedSize(horizontal: false, vertical: true)
                    .frame(maxWidth: .infinity, minHeight: Metrics.buttonHeight)
            } else {
                Button {} label: {
                    actionLabel(String(localized: "Preparing image…"), systemImage: nil, prominent: false)
                }
                .buttonStyle(.glass)
                .tint(Accent.graphite)
                .disabled(true)
            }
        }
        .padding(.horizontal, Metrics.hPadding)
        .padding(.bottom, 4)
        .frame(maxWidth: Metrics.maxContentWidth)
        .frame(maxWidth: .infinity)
    }

    /// `prominent` picks the ink: the INVERSE ink is correct only over
    /// `.glassProminent`'s opaque graphite fill — on plain glass it rendered white on
    /// the light slate field, the exact white-on-white trap CLAUDE.md records, one
    /// colour scheme over.
    private func actionLabel(_ title: String, systemImage: String?,
                             prominent: Bool) -> some View {
        HStack(spacing: 8) {
            if let systemImage { Image(systemName: systemImage) }
            Text(title)
        }
        .font(.system(.headline, weight: .semibold))
        .foregroundStyle(prominent
                         ? Color.adaptive(Color(hex: "FFFFFF"), Color(hex: "1B1F25"))
                         : Accent.graphite)
        .lineLimit(1)
        .minimumScaleFactor(0.85)
        // Full-width and padded, so the drawn label is nowhere near the tappable area
        // SwiftUI would infer from it on its own.
        .frame(maxWidth: .infinity, minHeight: Metrics.buttonHeight)
        .contentShape(.rect)
    }

    private var exportCard: RoutineShareExportCard {
        RoutineShareExportCard(name: request.name, payload: payload)
    }

    /// Rendered once, when the sheet appears — the request is frozen, so there is no
    /// input that could change under it. `isOpaque = false` keeps the card's rounded
    /// corners transparent instead of squaring the PNG off with white.
    @MainActor
    private func renderImage() async {
        // The QR must have drawn before the card is baked — a failed generation renders
        // the fallback SENTENCE, and exporting a white card whose only content is an
        // error message would hand a friend exactly that.
        guard QRCodeView.canRender(payload) else {
            renderFailed = true
            return
        }
        let renderer = ImageRenderer(content: exportCard)
        renderer.isOpaque = false
        renderer.scale = 3
        renderer.proposedSize = ProposedViewSize(
            width: RoutineShareExportCard.width,
            height: RoutineShareExportCard.height)
        guard let cgImage = renderer.cgImage else {
            renderFailed = true
            return
        }
        let png = await ShareImageEncoder.shared.pngData(for: cgImage)
        guard !Task.isCancelled else { return }
        guard let png else {
            renderFailed = true
            return
        }
        rendered = RenderedRoutineShare(
            image: UIImage(cgImage: cgImage, scale: 3, orientation: .up),
            file: RoutineSharePNG(data: png))
    }
}

private struct RenderedRoutineShare {
    let image: UIImage
    let file: RoutineSharePNG
}

/// A file-backed transfer keeps the `.png` extension as well as the content type — and
/// carries the renderer's ORIGINAL bytes, so nothing re-encodes the image on the way to
/// whichever app the share sheet hands it to.
private struct RoutineSharePNG: Transferable {
    let data: Data

    static var transferRepresentation: some TransferRepresentation {
        FileRepresentation(exportedContentType: .png) { item in
            let directory = FileManager.default.temporaryDirectory
                .appendingPathComponent(UUID().uuidString, isDirectory: true)
            try FileManager.default.createDirectory(
                at: directory,
                withIntermediateDirectories: true)
            let url = directory.appendingPathComponent("get-a-grip-routine.png")
            try item.data.write(to: url, options: .atomic)
            return SentTransferredFile(url)
        }
    }
}

/// The shared asset. Every colour here is FIXED rather than adaptive: `ImageRenderer`
/// resolves an adaptive colour against whatever trait collection it happens to inherit,
/// and a card whose ink flipped with the sharer's appearance setting would export dark
/// modules on a dark field. It is also the reason `EdgeMark` stays off this card — the
/// mark draws in `Accent.graphite`, which inverts.
private struct RoutineShareExportCard: View {
    static let width: CGFloat = 360
    static let height: CGFloat = 448

    let name: String
    let payload: String

    private static let ink = Color(hex: "1B1F25")

    var body: some View {
        VStack(spacing: 16) {
            QRCodeView(string: payload, size: 236)

            Text(name)
                .font(.system(.title3, weight: .semibold))
                .lineLimit(2)
                .multilineTextAlignment(.center)
                .minimumScaleFactor(0.6)

            // The image travels on its own, to someone who may never have heard of the
            // app. "GET A GRIP" alone names the sender, not the action.
            Text("iPhone: Camera. Android: Get a Grip → Today menu → Scan a routine.")
                .font(.system(.caption))
                .multilineTextAlignment(.center)
                .opacity(0.7)

            Spacer(minLength: 8)

            Text("GET A GRIP")
                .font(.labelCaps())
                .tracking(0.8)
                .opacity(0.7)
        }
        .foregroundStyle(Self.ink)
        .padding(22)
        .frame(width: Self.width, height: Self.height, alignment: .top)
        .background {
            RoundedRectangle(cornerRadius: 28, style: .continuous)
                .fill(Color.white)
        }
        .accessibilityHidden(true)
    }
}

#Preview {
    RoutineShareSheet(
        request: RoutineShareRequest(
            name: "Daily no-hangs",
            metaLine: "20 mm · 6 sets · 36 pulls · ≈21 min",
            signatureFingers: .four,
            peakIntensity: 0.2,
            url: URL(string: "getagrip://routine#preview-payload")!),
        onClose: {})
}
